include/sys_clock.h

```diff
@@ -28,6 +28,82 @@
 extern "C" {
 #endif
 
+/**
+ * @addtogroup clock_apis
+ * @{
+ */
+
+/**
+ * @brief Tick precision used in timeout APIs
+ *
+ * This type defines the word size of the timeout values used in
+ * k_timeout_t objects, and thus defines an upper bound on maximum
+ * timeout length (or equivalently minimum tick duration).  Note that
+ * this does not affect the size of the system uptime counter, which
+ * is always a 64 bit count of ticks.
+ */
+#ifdef CONFIG_TIMEOUT_64BIT
+typedef s64_t k_ticks_t;
+#else
+typedef u32_t k_ticks_t;
+#endif
+
+#define K_TICKS_FOREVER ((k_ticks_t) -1)
+
+#ifndef CONFIG_LEGACY_TIMEOUT_API
+
+typedef struct {
+	k_ticks_t ticks;
+} k_timeout_t;
+
+/**
+ * @brief Compare timeouts for equality
+ *
+ * The k_timeout_t object is an opaque struct that should not be
+ * inspected by application code.  This macro exists so that users can
+ * test timeout objects for equality with known constants
+ * (e.g. K_NO_WAIT and K_FOREVER) when implementing their own APIs in
+ * terms of Zephyr timeout constants.
+ *
+ * @return True if the timeout objects are identical
+ */
+#define K_TIMEOUT_EQ(a, b) ((a).ticks == (b).ticks)
+
+#define Z_TIMEOUT_NO_WAIT ((k_timeout_t) {})
+#define Z_TIMEOUT_TICKS(t) ((k_timeout_t) { .ticks = (t) })
+#define Z_FOREVER Z_TIMEOUT_TICKS(K_TICKS_FOREVER)
+#define Z_TIMEOUT_MS(t) Z_TIMEOUT_TICKS(k_ms_to_ticks_ceil32(MAX(t, 0)))
+#define Z_TIMEOUT_US(t) Z_TIMEOUT_TICKS(k_us_to_ticks_ceil32(MAX(t, 0)))
+#define Z_TIMEOUT_NS(t) Z_TIMEOUT_TICKS(k_ns_to_ticks_ceil32(MAX(t, 0)))
+#define Z_TIMEOUT_CYC(t) Z_TIMEOUT_TICKS(k_cyc_to_ticks_ceil32(MAX(t, 0)))
+
+/* Converts between absolute timeout expiration values (packed into
+ * the negative space below K_TICKS_FOREVER) and (non-negative) delta
+ * timeout values.  If the result of Z_TICK_ABS(t) is >= 0, then the
+ * value was an absolute timeout with the returend expiration time.
+ * Note that this macro is bidirectional: Z_TICK_ABS(Z_TICK_ABS(t)) ==
+ * t for all inputs, and that the representation of K_TICKS_FOREVER is
+ * the same value in both spaces!  Clever, huh?
+ */
+#define Z_TICK_ABS(t) (K_TICKS_FOREVER - 1 - (t))
```

Comment:
maybe something with `GET_ABS` in name. Otherwise it can be confused with converting t to absolute (positive) value.

Reply:
It's a little subtle in general, honestly, thus that comment.  I don't know that a simple name is going to really help that much -- this is an internal gadget used in like three places.